File: epi_judge_java_solutions/epi/DirectoryPathNormalization.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Iterator;

public class DirectoryPathNormalization {
  @EpiTest(testDataFile = "directory_path_normalization.tsv")

  public static String shortestEquivalentPath(String path) {

    if (path.equals("")) {
      throw new IllegalArgumentException("Empty string is not a legal path.");
    }

    Deque<String> pathNames = new ArrayDeque<>();
    // Special case: starts with "/", which is an absolute path.
    if (path.startsWith("/")) {
      pathNames.addFirst("/");
    }

    for (String token : path.split("/")) {
      if (token.equals("..")) {
        if (pathNames.isEmpty() || pathNames.peekFirst().equals("..")) {
          pathNames.addFirst(token);
        } else {
          if (pathNames.peekFirst().equals("/")) {
            throw new IllegalArgumentException(
                "Path error, trying to go up root " + path);
          }
          pathNames.removeFirst();
        }
      } else if (!token.equals(".") && !token.isEmpty()) { // Must be a name.
        pathNames.addFirst(token);
      }
    }

    StringBuilder result = new StringBuilder();
    if (!pathNames.isEmpty()) {
      Iterator<String> it = pathNames.descendingIterator();
      String prev = it.next();
      result.append(prev);
      while (it.hasNext()) {
        if (!prev.equals("/")) {
          result.append("/");
        }
        prev = it.next();
        result.append(prev);
      }
    }
    return result.toString();
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "DirectoryPathNormalization.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_cpp_solutions/directory_path_normalization.cc
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

#include "test_framework/generic_test.h"

using std::invalid_argument;
using std::string;
using std::stringstream;
using std::vector;

string ShortestEquivalentPath(const string& path) {
  if (empty(path)) {
    throw invalid_argument("Empty string is not a valid path.");
  }

  vector<string> path_names;  // Uses vector as a stack.
  // Special case: starts with "/", which is an absolute path.
  if (path.front() == '/') {
    path_names.emplace_back("/");
  }

  stringstream ss(path);
  string token;
  while (getline(ss, token, '/')) {
    if (token == "..") {
      if (empty(path_names) || path_names.back() == "..") {
        path_names.emplace_back(token);
      } else {
        if (path_names.back() == "/") {
          throw invalid_argument("Path error");
        }
        path_names.pop_back();
      }
    } else if (token != "." && token != "") {  // Must be a name.
      path_names.emplace_back(token);
    }
  }

  string result;
  if (!empty(path_names)) {
    result = path_names.front();
    for (int i = 1; i < size(path_names); ++i) {
      if (i == 1 && result == "/") {  // Avoid starting "//".
        result += path_names[i];
      } else {
        result += "/" + path_names[i];
      }
    }
  }
  return result;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"path"};
  return GenericTestMain(args, "directory_path_normalization.cc",
                         "directory_path_normalization.tsv",
                         &ShortestEquivalentPath, DefaultComparator{},
                         param_names);
}
